controller/snapshot_controller.go

```diff
@@ -344,9 +347,19 @@ func (sc *SnapshotController) reconcile(snapshotName string) (err error) {
 			return sc.handleAttachmentTicketDeletion(snapshot)
 		}
 
-		if err := sc.handleAttachmentTicketCreation(snapshot); err != nil {
+		// It is possible for us to be reconciling an out-of-date snapshot that has had its finalizer removed and has
+		// been deleted from the cluster, so we must take special care around attachment tickets. It is fine to update
+		// an attachment ticket after DeletionTimestamp is set (because attempts to update an out-of-date volume
+		// attachment will simply fail), but we must only create one once (to avoid recreating one the list time we
+		// reconcile).
+		requireVATicket := snapshot.Annotations[types.AttachingForDeletionAnnotationKey] == "true"
+		if err := sc.handleAttachmentTicketModification(snapshot, requireVATicket); err != nil {
 			return err
 		}
+		if snapshot.Annotations == nil {
+			snapshot.Annotations = make(map[string]string)
+		}
+		snapshot.Annotations[types.AttachingForDeletionAnnotationKey] = "true"
```

Comment:
Interesting. Yes, I think you are right. My original logic included a return if the annotation was not set. We would NEVER proceed to delete the finalizer if we weren't operating on an annotated snapshot, so there was no way of this happening.

https://github.com/longhorn/longhorn-manager/pull/2147/commits/10b343eced295276fe3467d287c6ee89a719417d

I removed the immediate return while simplifying things during the review. IIRC, the new code still passed my testing, but we need to put the immediate return back. Unfortunately, I'm not going to get this done before I leave for vacation, but WDYT of the flow in that linked commit?

Reply:
Sounds good. Let's continue on this one after you are back from the vacation.
The logic in the liked comment LGTM.

Reply:
Btw, I see another potential root cause that can trigger the issue in one sync loop:
1. snapshot controller create VA ticket https://github.dev/longhorn/longhorn-manager/blob/714bc966d5e271e9f1159b9266578da7f8a27600/controller/snapshot_controller.go#L347-L349
1. The VA object in ETCD is updated but not in the cache
1. snapshot controller checks the VA object in the cache and doesn't delete/update the VA because it doesn't see the ticket 
1. snapshot controller removes finalizer.
1. Assume that cach is updated fast enough so that in the next sync loop, snapshot controller doesn't see snapshot CR and never sync it again
1. Now we have orphan VA ticket

Reply:
It seems we have a concern to use outdated cache to do the resource reconciling and following update. Should we just bypass the cache to do resource query and update? Only use the cache informer to watch resource change. 

Can we check the best practice of controller implementation as well? like K8s builtin controllers, have we checked them?